Mutagen.Bethesda.Analyzers.Skyrim/Record/Armor/RaceAddonsAnalyzer.cs

```diff
@@ -0,0 +1,75 @@
+            "Armor is missing race addons",
+            Severity.Warning)
+        .WithFormatting<string?>("Missing race addon for race: {0}");
+
+    public IEnumerable<TopicDefinition> Topics => [ArmorMissingRaceAddons];
+
+    private static readonly List<FormKey> DefaultPlayerRaces =
+    [
+        FormKeys.SkyrimSE.Skyrim.Race.ArgonianRace.FormKey, FormKeys.SkyrimSE.Skyrim.Race.ArgonianRaceVampire.FormKey,
+        FormKeys.SkyrimSE.Skyrim.Race.BretonRace.FormKey, FormKeys.SkyrimSE.Skyrim.Race.BretonRaceVampire.FormKey,
+        FormKeys.SkyrimSE.Skyrim.Race.DarkElfRace.FormKey, FormKeys.SkyrimSE.Skyrim.Race.DarkElfRaceVampire.FormKey,
+        FormKeys.SkyrimSE.Skyrim.Race.HighElfRace.FormKey, FormKeys.SkyrimSE.Skyrim.Race.HighElfRaceVampire.FormKey,
+        FormKeys.SkyrimSE.Skyrim.Race.ImperialRace.FormKey, FormKeys.SkyrimSE.Skyrim.Race.ImperialRaceVampire.FormKey,
+        FormKeys.SkyrimSE.Skyrim.Race.KhajiitRace.FormKey, FormKeys.SkyrimSE.Skyrim.Race.KhajiitRaceVampire.FormKey,
+        FormKeys.SkyrimSE.Skyrim.Race.NordRace.FormKey, FormKeys.SkyrimSE.Skyrim.Race.NordRaceVampire.FormKey,
+        FormKeys.SkyrimSE.Skyrim.Race.OrcRace.FormKey, FormKeys.SkyrimSE.Skyrim.Race.OrcRaceVampire.FormKey,
+        FormKeys.SkyrimSE.Skyrim.Race.RedguardRace.FormKey, FormKeys.SkyrimSE.Skyrim.Race.RedguardRaceVampire.FormKey,
+        FormKeys.SkyrimSE.Skyrim.Race.WoodElfRace.FormKey, FormKeys.SkyrimSE.Skyrim.Race.WoodElfRaceVampire.FormKey
+    ];
+
+    public RecordAnalyzerResult? AnalyzeRecord(ContextualRecordAnalyzerParams<IArmorGetter> param)
+    {
+        var armor = param.Record;
+
+        // Armor with template armor inherit all relevant data from the template armor and should not be checked themselves
+        if (!armor.TemplateArmor.IsNull) return null;
+
+        // Non-playable cannot be equipped by the player and is usually required to be universally compatible
+        if (armor.MajorFlags.HasFlag(Bethesda.Skyrim.Armor.MajorFlag.NonPlayable)) return null;
+
+        // Exclude non player race armor
+        if (armor.Race.FormKey != FormKeys.SkyrimSE.Skyrim.Race.DefaultRace.FormKey
+            && !DefaultPlayerRaces.Contains(armor.Race.FormKey)) return null;
+
+        // Build list of all races in armor addons
+        var missingRaces = new List<FormKey>(DefaultPlayerRaces);
+        foreach (var armorAddon in armor.Armature)
+        {
+            var addon = armorAddon.TryResolve(param.LinkCache);
+            if (addon == null) continue;
+
+            foreach (var additionalRace in addon.AdditionalRaces)
+            {
+                missingRaces.Remove(additionalRace.FormKey);
+            }
+            missingRaces.Remove(addon.Race.FormKey);
+        }
+
+        var result = new RecordAnalyzerResult();
+
+        foreach (var race in missingRaces)
+        {
+            if (param.LinkCache.TryResolve<IRaceGetter>(race, out var r))
```

Comment:
Hmm.  The resolution here to get the EditorID feels wrong, as the editor ID is just for "convenience" in the log message to include the EditorID in addition to the FormKey?

The problem is that if the Race cant be found, this topic wont be added.  I'm guessing that's not desired?

I think instead, we'll want to upgrade the engine.   We should be able to provide a topic here with just the FormKey of the race.  The engine can do the lookup and decorate with the EditorID when appropriate?  That way the topic gets reported either way, but the EditorID is included when it can be, and also each analyzer doesn't need to remember to do anything special to get the perk

https://github.com/Mutagen-Modding/Mutagen.Bethesda.Analyzers/issues/123

Reply:
Yeah I agree that would be good